packages/iTwinUI-react/src/core/Dialog/Dialog.tsx

```diff
@@ -0,0 +1,88 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+import cx from 'classnames';
+import { useTheme } from '../utils';
+import '@itwin/itwinui-css/css/dialog.css';
+import { DialogTitleBar } from './DialogTitleBar';
+import { DialogContent } from './DialogContent';
+import { CSSTransition } from 'react-transition-group';
+
+export type DialogProps = {
+  /**
+   * Flag whether dialog should be shown.
+   * @default false
+   */
+  isOpen?: boolean;
+  /**
+   * Type of the dialog.
+   * @default 'default'
+   */
+  styleType?: 'default' | 'fullPage';
+  /**
+   * Content of the dialog.
+   */
+  children: React.ReactNode;
+} & React.ComponentPropsWithRef<'div'>;
+
+/**
+ * Dialog component which can wrap any content.
+ * @example
+ * <Dialog>
+ *   <Dialog.TitleBar>
+ *     My dialog title
+ *   </Dialog.TitleBar>
+ *   <Dialog.Content>
+ *     Here is my dialog content
+ *   </Dialog.Content>
+ *   <Dialog.ButtonBar>
+ *     <Button styleType='high-visibility'>
+ *       Primary button
+ *     </Button>
+ *     <Button>
+ *       Secondary button
+ *     </Button>
+ *   </Dialog.ButtonBar>
+ * </Dialog>
+ */
+export const Dialog = Object.assign(
+  React.forwardRef<HTMLDivElement, DialogProps>((props, ref) => {
+    const { className, children, styleType, isOpen = false, ...rest } = props;
+
+    useTheme();
+
+    return (
+      <CSSTransition
```

Comment:
I am not sure whether to have `CSSTransition` here or not.
Pros: animations come straight out of box.
Cons: maybe it makes `Dialog` not so dumb and reusable component.

Reply:
Didn't we talk about adding a wrapper component? Maybe lets move all the extra logic there?
![image](https://user-images.githubusercontent.com/9084735/173859338-ae10a200-97d8-42f5-a246-7d6ee3ec2ab0.png)

Reply:
Added a wrapper with a context. I kind of copied from radix-ui. But I am still not sure about that.

Reply:
It looks like backdrop still has a lot of logic. Would be nice to somehow move all that out of it.

What if we add a `backdrop` prop which accepts the generic `<Backdrop />` component? Then we can handle all the logic in Dialog?

Reply:
But then we would have some kind of mix between composition and passing components trough props. Also, what it gives moving logic from `DialogBackdrop` to 'Dialog`?

Reply:
I think passing components through props is still somewhat like composition, as it gives the user control over the actual component that gets rendered.

The way I see it, backdrop needs to be a super simple component and we should try to handle as much as logic as we can in the dialog itself. This way, we could even avoid creating a special `Dialog.Backdrop` component and just reuse `Backdrop` (or the user could pass their own).

Plus, I think some of the existing things like keydown handlers and focus should already be moved to the dialog body anyway. I find it really odd that our focus gets "lost" on the backdrop (really hard to see visually). The only thing that where we might need to mess with backdrop is the "outside click", but I think we could even handle that by adding events on the `document`.

Another benefit of this approach would be that it would make it more compatible with the native `<dialog>` element when we decide to support it.

Reply:
https://github.com/iTwin/iTwinUI-react/pull/670#discussion_r920835897